### PR TITLE
Esteban/integrate users with categories

### PR DESCRIPTION
**Pull request by esteban-gs**
## Dev Summary

This PR proposes to store the users' interests in an `interests` field of type `JSONB` instead of adding a new many-to-many table. The trade-off is that if a specific `categories` record changes, the list of interests won't change automatically. The assumption is that the user will be able to update their interests in their profiles as they wish. In that case, a new interest list can be stored in the `interests`  field.

The PR also includes
- a working service method to query the users based on a list of interests
- a new method in the categories service to validate a list of categories
- 2 new custom dto validators to validate the correct shape of the `JSONB` interests object

## Test Plan

repro steps:

1. run the app.
2. register a new user
```bash
curl --location 'http://localhost:3001/api/auth/register' \
--header 'Content-Type: application/json' \
--data-raw '{
    "firstName": "Billie",
    "last_name": "Eilish",
    "email": "user5@example.com",
    "password": "Secret1234%",
    "email_notification_opt_out": true,
    "city": "seattle",
    "state": "wa",
    "interests": {
        "names": [
            "furniture"
        ]
    }
}'
```
3. log in and expect to see a response that includes the the interests field:
```json
.....
        "interests": {
            "names": [
                "furniture"
            ]
        }
.....
```
4. validate categories endpoint should return true when the categories are valid, otherwise, false
```bash
curl --location 'http://localhost:3001/api/categories/validate' \
--header 'Content-Type: application/json' \
--header 'Cookie: NEH_is_cool={your_cookie}' \
--data '[
    "paper products",
    "furniture"
]'
```

## Resources

- [postgres JSONB with typeorm](https://wanago.io/2020/12/28/nestjs-json-postgresql-typeorm/)

### CODE REVIEW

**Comment by garimacha** [comment]
Is anyone going to review this ? @jd2rogers2  @brodylosh @rmwerner18 @EtoKruto

**Comment by esteban-gs** [comment]
> Is anyone going to review this ? @jd2rogers2 @brodylosh @rmwerner18 @EtoKruto

Thanks for checking in @garimacha , I tagged the engineers here for visibility, mostly.

**Review by rmwerner18 — APPROVED** [review]

**Review by jd2rogers2 — COMMENTED** [review]
_i think_ i see some improvements, but you're closer to the code so if i'm wrong go with your solution

File: server/src/acccount-manager/account-manager.controller.ts [hunk 1]
```diff
@@ -62,9 +62,13 @@ export class AccountManagerController {
   }
 
   @Post('register')
-  async register(
-    @Body() createUserDto: CreateUserDto,
-  ): Promise<Omit<User, 'password' | 'accept_terms'>> {
+  async register(@Body() createUserDto: CreateUserDto) {
```
[the remark this answers is not in the document]
Comment (esteban-gs) [reply server/src/acccount-manager/account-manager.controller.ts:65]: 🤦 I removed it inadvertently

File: server/src/acccount-manager/account-manager.controller.ts [hunk 2]
```diff
@@ -62,9 +62,13 @@ export class AccountManagerController {
...
+    if (createUserDto.interests) {
+      const res = await this.accountManagerService.validateInterests(createUserDto.interests.names);
```
[the remark this answers is not in the document]
Comment (esteban-gs) [reply server/src/acccount-manager/account-manager.controller.ts:67]: Mainly to keep the dependencies of the `AccountManagerController.ts` and limit its responsibility to http things.

File: server/src/categories/categories.service.ts [hunk 3]
```diff
@@ -23,6 +23,13 @@ export class CategoriesService {
     return this.categoriesRepository.find({ where: getCategoriesDto });
   }
 
+  async validateCategories(categories: string[]): Promise<boolean> {
```
[the remark this answers is not in the document]
Comment (esteban-gs) [reply server/src/categories/categories.service.ts:26]: This is a great point and I think that with this set up we could move to enums.

[the remark this answers is not in the document]
Comment (jd2rogers2) [reply server/src/categories/categories.service.ts:26]: chatted offline and the table w/ custom validator is the best option

**Review by jd2rogers2 — APPROVED** [review]

**Pull request merged**